proposals/csharp-12.0/collection-expressions.md

```diff
@@ -219,7 +213,79 @@ ImmutableArray<int> ia =
 ## Construction
 [construction]: #construction
 
-*Give specific ordering for determining how to construct the constructible collection types.*
+The elements of a collection expression are *evaluated* in order, left to right.
+Each element is evaluated exactly once, and any further references to the elements refer to the results of this initial evaluation.
+
+Spread elements are *iterated* in the order the elements appear in the collection expression, left to right. A spread element may be *iterated* before or after the subsequent elements in the collection expression are *evaluated*.
+
+An unhandled exception thrown from any of the methods used during construction will be uncaught and will prevent further steps in the construction.
+
+`Length` and `Count` properties, and `GetEnumerator` methods, are assumed to have no side effects.
+
+---
+
+If the target type is a *struct* or *class type* that implements `System.Collections.IEnumerable`, and the target type does not have a *[create method](#create-methods)*, the construction of the collection instance is as follows:
+
+* The elements are evaluated in order. Some or all elements may be evaluated *during* the steps below rather than before.
+
+* The compiler *may* determine the *known length* of the collection expression by invoking [*countable*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md#adding-index-and-range-support-to-existing-library-types) properties &mdash; or equivalent properties from well-known interfaces or types &mdash; on each *spread element expression*.
```

Comment:
this doesn't preclude us using TryGetNonEnumeratedCount in the future right?

Reply:
It shouldn't and I didn't intend it to.